### PR TITLE
[BE][Attention] Factor out common code

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139791
* #139763
* #139784
* __->__ #139788

- Compute attention mask before the switch
- Introduce `query_device_type` variable
- Refactor some of MPS-math checks into easily readable boolean names

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

File: aten/src/ATen/native/transformers/attention.cpp [hunk 1]
```diff
@@ -703,30 +703,32 @@ Tensor scaled_dot_product_attention(
     bool is_causal,
     std::optional<double> scale,
     bool enable_gqa) {
+  using namespace sdp;
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/attention.cpp:706]: personally dont love maybe just bring in the two objects in the namespace

File: aten/src/ATen/native/transformers/attention.cpp [hunk 2]
```diff
@@ -745,18 +746,19 @@ Tensor scaled_dot_product_attention(
...
       return std::get<0>(out_lse_softmax);
     }
-    case sdp::SDPBackend::math: {
-      std::optional<Tensor> attn_mask = convert_boolean_attn_mask(attn_mask_, query_.dtype());
-      if ((!GradMode::is_enabled() || (!query_.requires_grad() && !key.requires_grad() && !value.requires_grad()))
-          && query_.device().type() == DeviceType::MPS && dropout_p == 0.0
-          && query_.is_contiguous() && key.is_contiguous() && value.is_contiguous()
-          && !query_.is_nested() && !key.is_nested() && !value.is_nested()) {
+    case SDPBackend::math: {
+      const auto any_nested = query_.is_nested() || key.is_nested() || value.is_nested();
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/attention.cpp:755]: So these checks should ideally be in the dispatching logic, I am not sure if MPS currently has a stub for doing the input validation. I think that it this length of check its okay but if it gets any larger I think we should do it properly

**Review by drisspg — APPROVED** [review]
